python/zarrs/pipeline.py

```diff
@@ -44,6 +50,55 @@ def get_codec_pipeline_impl(codec_metadata_json: str) -> CodecPipelineImpl:
     )
 
 
+def codecs_to_dict(codecs: Iterable[Codec]) -> Generator[dict[str, Any], None, None]:
+    for codec in codecs:
+        if codec.__class__.__name__ == "V2Codec":
+            codec_dict = codec.to_dict()
+            compressor = {}
+            if codec_dict.get("compressor", None) is not None:
+                compressor = codec_dict["compressor"].get_config()
+                if compressor.get("id", None) == "zstd":
+                    yield {
+                        "name": "zstd",
+                        "configuration": {
+                            "level": int(compressor["level"]),
+                            "checksum": compressor["checksum"],
+                        },
+                    }
+                elif compressor.get("id", None) == "blosc":
+                    as_dict = {
+                        "name": "blosc",
+                        "configuration": {
+                            "cname": compressor["cname"],
+                            "clevel": int(compressor["clevel"]),
+                            "blocksize": int(compressor["blocksize"]),
+                        },
+                    }
+                    if typesize := compressor.get("typesize", None) is not None:
+                        as_dict["typesize"] = typesize
+                    if shuffle := compressor.get("shuffler", None) is not None:
+                        # https://github.com/LDeakin/zarrs/blob/0532fe983b7b42b59dbf84e50a2fe5e6f7bad4ce/zarrs_metadata/src/v3/array/codec/blosc.rs#L46-L54
+                        match shuffle:
+                            case 0:
+                                as_dict["shuffle"] = "noshuffle"
+                            case 1:
+                                as_dict["shuffle"] = "shuffle"
+                            case 2:
+                                as_dict["shuffle"] = "bitshuffle"
+                    yield as_dict
+                else:
+                    yield compressor
+            elif codec_dict.get("filter", None) is not None:
+                filter_ = codec_dict["filter"].get_config()
+                yield filter_
+            # TODO: get the endianness added to V2Codec API
+            # TODO: how to handle this with strings, which don't need this but zarrs
+            # complains about its absence if its not there
+            yield BytesCodec().to_dict()
```

Comment:
This seems to be both necessary and causing issues - without it, there is a lot of complaining about no array-bytes codecs, but with it, string i/o is broken complaining about no-compatibility with this codec